client/importServiceWorker.js

```diff
@@ -1,18 +1,77 @@
 import { Meteor } from 'meteor/meteor';
+import { Tracker } from 'meteor/tracker';
 
-Meteor.startup(() => {
+import { settings } from '../app/settings';
+import { handleError, t } from '../app/utils/client';
+import { modal } from '../app/ui-utils/client';
+
+function urlBase64ToUint8Array(base64String) {
+	const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
+	const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');
+
+	const rawData = atob(base64);
+	const outputArray = new Uint8Array(rawData.length);
+
+	for (let i = 0; i < rawData.length; ++i) {
+		outputArray[i] = rawData.charCodeAt(i);
+	}
+	return outputArray;
+}
+
+function subscribeUser() {
 	if ('serviceWorker' in navigator) {
-	// if (navigator.serviceWorker.controller) {
-	//	console.log('Active service worker found, no need to register');
-	// } else {
-		// Register the service worker
-		navigator.serviceWorker
-			.register('sw.js', {
-				scope: './',
-			})
-			.then(function(reg) {
-				console.log(`Service worker has been registered for scope: ${ reg.scope }`);
-			});
-	// }
+		navigator.serviceWorker.ready.then(async function(reg) {
+			try {
+				const vapidKey = await settings.get('Vapid_public_key');
+				const subscription = await reg.pushManager
+					.subscribe({
+						userVisibleOnly: true,
+						applicationServerKey: urlBase64ToUint8Array(vapidKey),
+					});
+				Meteor.call('savePushNotificationSubscription', JSON.stringify(subscription));
+			} catch (e) {
+				handleError(e);
+			}
+		});
 	}
+}
+
+Meteor.startup(() => {
+	Tracker.autorun((computation) => {
+		const settingsReady = settings.cachedCollection.ready.get();
+		if ('serviceWorker' in navigator) {
+			navigator.serviceWorker
+				.register('sw.js', {
+					scope: './',
+				})
+				.then(function(reg) {
+					reg.pushManager.getSubscription().then(function(sub) {
+						if (sub === null) {
+							console.log('Not subscribed to push service!');
+							if (settingsReady) {
+								modal.open({
+									title: t('Important'),
+									type: 'info',
+									text: t('Please subscribe to push notifications to continue'),
+									showCancelButton: true,
+									confirmButtonText: t('Subscribe'),
+									cancelButtonText: t('Cancel'),
+									closeOnConfirm: true,
+								}, () => {
+									Notification.requestPermission().then(function(permission) {
+										if (permission === 'granted') {
+											subscribeUser();
+										}
+									});
+								});
+								computation.stop();
+							}
```

Comment:
Do we need a retry mechanism over here?

Reply:
This is just to wait for the settings to load. Till the time settings are not loaded this won't be triggered. It will be triggered only once. We take the application key from the settings so we need to wait for that